rueidisotel/trace.go

```diff
@@ -27,12 +27,26 @@ var (
 
 var _ rueidis.Client = (*otelclient)(nil)
 
-// WithClient creates a new rueidis.Client with OpenTelemetry tracing enabled
+// WithClient creates a new rueidis.Client with OpenTelemetry tracing enabled.
 func WithClient(client rueidis.Client, opts ...Option) rueidis.Client {
-	o := &otelclient{client: client}
-	for _, fn := range opts {
-		fn(o)
+	o := &otelclient{
+		client: client,
 	}
+	for _, opt := range opts {
+		opt(o)
+	}
+	if o.meterProvider == nil {
+		o.meterProvider = otel.GetMeterProvider() // Default to global MeterProvider
+	}
+	if o.tracerProvider == nil {
+		o.tracerProvider = otel.GetTracerProvider() // Default to global TracerProvider
+	}
+	// Now that we have the meterProvider and tracerProvider, get the Meter and Tracer
+	o.meter = o.meterProvider.Meter(name)
+	o.tracer = o.tracerProvider.Tracer(name)
+	// Now create the counters using the meter
+	o.cscMiss, _ = o.meter.Int64Counter("rueidis_do_cache_miss")
+	o.cscHits, _ = o.meter.Int64Counter("rueidis_do_cache_hits")
```

Comment:
So, after moving the tracer and counters to here, the original global ones should be removed.

Reply:
got it

Reply:
Hi @HoganEdwardChu,

I just looked into these tests. They all failed with nil dereferencing on the `tracer` field. So you need to add them to all `&otelclient{...}` and `&dedicated{...}`

Reply:
Hi @rueian  i added it in https://github.com/redis/rueidis/blob/d29aa821a42d9a5affbe0a2217bdaa99fe715a5c/rueidisotel/trace.go#L41 and https://github.com/redis/rueidis/blob/d29aa821a42d9a5affbe0a2217bdaa99fe715a5c/rueidisotel/trace.go#L85 you mean https://github.com/redis/rueidis/blob/d29aa821a42d9a5affbe0a2217bdaa99fe715a5c/rueidisotel/trace.go#L171 here?